src/algorithms/reco/PrimaryVertices.h

```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2024 Daniel Brandenburg
+
+#pragma once
+
+#include <edm4eic/VertexCollection.h>
+#include <spdlog/logger.h>
+#include <memory>
+
+#include "algorithms/reco/PrimaryVerticesConfig.h"
+#include "algorithms/interfaces/WithPodConfig.h"
+
+
+namespace eicrecon {
+
+    class PrimaryVertices : public WithPodConfig<PrimaryVerticesConfig>{
+
+    public:
+
+        void init(std::shared_ptr<spdlog::logger>& logger);
+        std::unique_ptr<edm4eic::VertexCollection> execute(
+                const edm4eic::VertexCollection *rcvtx
+        );
+
+    private:
+        std::shared_ptr<spdlog::logger> m_log;
```

Comment:
This is not how we want algorithms to be defined.

1. Have to `#include <algorithms/algorithm.h>`
2. Define your algorithm type using algorithms::Algorithm
3. `void init()` has to be `final` and not take any arguments (use services for logger)
4. `void process(const Input&, const Output&) const final;` is the signature and name of the main entry point, it does not return a value other than by reference `Output`.

See any other algorithm in EICrecon for a good example.

Reply:
Hi Dmitry,

I have seen the example from other EICrecon part, but do you know any one of them is dealing with the SubsetCollection? I tried the setup as you suggested, the codes compile, but crashes when tested with some errors. The difference I can see is that I was following the ScatteredElectron or ElectronReconstruction algorithm which also use SubsetCollection. Any suggestions on this? Thanks

Reply:
Hard to tell without seeing the code. An issue might be because you allocate a collection on your end, like you do on line 47 above.

Reply:
Submitted a conversion of the original code https://github.com/eic/EICrecon/pull/1615

Reply:
Updated following the instruction to use the algorithms interface. The issue a couple of days ago was due to the null logger pointer (fixed now). Please take a review of the new codes.